model/riscv_insts_cmoext.sail

```diff
@@ -0,0 +1,245 @@
+                (E_Extension(e), _ , _ ) => { ext_handle_data_check_error(e); RETIRE_FAIL }
+              }
+          }
+        }
+      }
+    }
+  }
+}
+
+function check_clean_flush(p : Privilege) -> bool = {
+  let ret : bool = ~(((p != Machine) & (menvcfg.CBCFE() == 0b0)) |
+                    ((p == User) & (senvcfg.CBCFE() == 0b0)));
+  ret
+}
+
+function check_inval(p : Privilege) -> bits(2) = {
+  let bit0 : bool = ((p != Machine) & (menvcfg.CBIE() == 0b00)) |
+                    ((p == User) & (senvcfg.CBIE() == 0b00));
+  let bit1 : bool = ((p != Machine) & (menvcfg.CBIE() == 0b01)) |
+                    ((p == User) & (senvcfg.CBIE() == 0b01));
+  if bit0 then EXTZ(0b0)
+  else if bit1 then EXTZ(0b1)
+  else EXTZ(0b11)
+}
+
+function check_zero(p : Privilege) -> bool = {
+  let ret : bool = ~(((p != Machine) & (menvcfg.CBZE() == 0b0)) |
+                    ((p == User) & (senvcfg.CBZE() == 0b0)));
+  ret
+}
+
+function clause execute(RISCV_CBO(cbop, rs1)) = {
+  let rs1_val = X(rs1);
+  let cache_block_size : xlenbits = plat_cache_block_size() ;
+  let vaddr: xlenbits = rs1_val & ~(cache_block_size - EXTZ(0b1));
+  match cbop {
+    CBO_CLEAN => {
+      if (check_clean_flush(cur_privilege)) then
+        /* perform clean*/
+        process_clean_inval(vaddr, cache_block_size, true, false)
+      else {
+        handle_illegal();
+        RETIRE_FAIL
+      }
+    },
+    CBO_FLUSH => {
+      if (check_clean_flush(cur_privilege)) then
+        /* perform flush*/
+        process_clean_inval(vaddr, cache_block_size, true, true)
+      else {
+        handle_illegal();
+        RETIRE_FAIL
+      }
+    },
+    CBO_INVAL => {
+      match (check_inval(cur_privilege)) {
+        0b00 => {
+          handle_illegal();
+          RETIRE_FAIL
+          },
+        0b01 => process_clean_inval(vaddr, cache_block_size, true, true),  /*perform flush*/
+        0b10 => process_clean_inval(vaddr, cache_block_size, false, true) /*perform inval*/
+      }
+    },
+    CBO_ZERO  => {
+      if (check_zero(cur_privilege)) then
+        process_cbo_zero(vaddr, cache_block_size)
+      else {
+        handle_illegal();
+        RETIRE_FAIL
+      }
+    }
+  }
+}
+
+union clause ast = RISCV_PREFETCH : (preop, regidx, bits(7))
+
+mapping encdec_prefetch : preop <-> bits(5) = {
+  PREFETCH_I <-> 0b00000,
+  PREFETCH_R <-> 0b00001,
+  PREFETCH_W <-> 0b00011
+}
+
+mapping clause encdec = RISCV_PREFETCH(op, rs1, offset)
+  <-> offset @ encdec_prefetch(op) @ rs1 @ 0b110 @ 0b00000 @ 0b0010011
+
+mapping pre_mnemonic : preop <-> string = {
+  PREFETCH_I <-> "prefetch.i",
+  PREFETCH_R <-> "prefetch.r",
+  PREFETCH_W <-> "prefetch.w"
+}
+
+mapping clause assembly = RISCV_PREFETCH(op, rs1, offset)
+  <-> pre_mnemonic(op) ^ spc() ^ hex_bits_12(offset @ 0b00000) ^ "(" ^ reg_name(rs1) ^ ")"
+
+function clause execute(RISCV_PREFETCH(preop, rs1, offset)) = {
+  /* do nothing other than translate_address for prefetch.i/r/w */
```

Comment:
Why do we even bother doing that? They're just hints, they can be entirely nop'ed out in the model like any other hint. Which is already the case, given they're ORI instructions, and base.sail gets included before cmoext.sail so the real ORIs will take precedence and this code is entirely dead.

Especially since this is actually wrong, due to translateAddr having the side-effect of (if enabled instead of trapping for software emulation) atomically updating accessed and dirty bits as needed, which is specifically not performed for prefetch.